src/devices/bus/nscsi/tape.cpp

```diff
@@ -0,0 +1,1028 @@
+				const bool sdp = m_pl_buf[page_pos + 4] & 0x40; // "select data partitions"
+				const bool idp = m_pl_buf[page_pos + 4] & 0x20; // "initiator-defined partitions"
+				const int psum = (m_pl_buf[page_pos + 4] & 0x18) >> 3; // "partition size unit of measure"
+				const int mfr = m_pl_buf[page_pos + 5]; // "medium format recognition"
+				LOG("        map=%d apd=%d fdp=%d sdp=%d idp=%d psum=0x%x mfr=0x%02x\n", map, apd, fdp, sdp, idp, psum, mfr);
+				int psd_num = 0; // partition size descriptor number
+				int psd_pos = 8; // partition size descriptor position
+				while (psd_pos < page_len) {
+					const int ps = get_uint16be(&m_pl_buf[page_pos + psd_pos]); // "partition size"
+					LOG("        psd_num=%d ps=%d\n", psd_num, ps);
+					psd_num++;
+					psd_pos += 2;
+				}
+				break;
+			}
+			default:
+				LOG("        *** BAD PAGE\n");
+		}
+		page_num++;
+		page_pos += 2 + page_len;
+	}
+	scsi_status_complete(SS_GOOD);
+}
+
+void nscsi_tape_device::handle_mode_sense_6() // mandatory; SCSI-2 sections 8.2.10, 8.3.3, 10.3.3
+{
+	const bool bd_disable = scsi_cmdbuf[1] & 0x80; // should we not respond with block descriptor
+	const int page_control = scsi_cmdbuf[2] >> 6;
+	const int page_code = scsi_cmdbuf[2] & 0x3f;
+	LOG("MODE SENSE(6) bd_disable=%d page_control=0x%02x page_code=0x%02x\n", bd_disable, page_control, page_code);
+	if (scsi_cmdbuf[1] & 0x17 || scsi_cmdbuf[3]) // error: reserved bits set
+		return report_bad_cdb_field();
+
+	if (page_code && page_code != SPC_RETURN_ALL_MODE_PAGES) // error: we don't support pages other than 0
+		return report_bad_cdb_field();
+
+	// TODO: support more pages, such as SPC_DATA_COMPRESSION_PAGE, and SPC_DEVICE_CONFIGURATION_PAGE with BIS bit set to indicate we support reporting block addresses in response to READ POSITION
+	if (page_control == SPC_SAVED_VALUES) // error: we don't support saving parameters
+		return report_no_saving_params();
+
+	const int resp_len = bd_disable ? 4 : 12; // response length
+	assert(sizeof(scsi_cmdbuf) >= resp_len);
+	memset(scsi_cmdbuf, 0, resp_len);
+	scsi_cmdbuf[0] = resp_len - 1; // mode data length
+	scsi_cmdbuf[2] = (m_image->get_file() && m_image->get_file()->is_read_only()) ? 0x80 : 0; // device-specific parameter
+	scsi_cmdbuf[3] = bd_disable ? 0 : 8; // block descriptor length
+	if (!bd_disable) { // respond with block descriptor
+		scsi_cmdbuf[4] = m_image->get_file() ? m_image->get_file()->get_density_code() : 0; // density code
+		put_uint24be(&scsi_cmdbuf[9], m_fixed_block_len); // block length
+	}
+	scsi_data_in(SBUF_MAIN, resp_len);
+	scsi_status_complete(SS_GOOD);
+}
+
+void nscsi_tape_device::handle_request_sense(const int lun) // mandatory; SCSI-2 section 8.2.14
+{
+	const int alloc_len = scsi_cmdbuf[4]; // allocation length
+	LOG("REQUEST SENSE lun=%d alloc_len=%d\n", lun, alloc_len);
+	if (scsi_cmdbuf[1] & 0x1f || scsi_cmdbuf[2] || scsi_cmdbuf[3]) // error: reserved bits set
```

Comment:
Please add parentheses around the binary and; although not required, some compilers will produce warnings here.